File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/autonomous/AutonomousOpMode.java
package org.firstinspires.ftc.teamcode.opmodes.autonomous;

import com.qualcomm.hardware.bosch.BNO055IMU;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.util.Range;

import org.firstinspires.ftc.robotcore.external.Func;
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
import org.firstinspires.ftc.robotcore.external.navigation.AxesOrder;
import org.firstinspires.ftc.robotcore.external.navigation.AxesReference;
import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit;
import org.firstinspires.ftc.robotcore.external.navigation.Orientation;
import org.firstinspires.ftc.robotcore.external.navigation.Position;
import org.firstinspires.ftc.robotcore.external.navigation.Velocity;
import org.firstinspires.ftc.teamcode.hardware.HardwareSkybot;
import org.firstinspires.ftc.teamcode.utilities.PIDController;

import java.util.Locale;

public abstract class AutonomousOpMode extends LinearOpMode {
    private final double aConstantProp = 0.000163;
    private final double bConstantProp = 0.009567;
    private final double cConstantProp = 0.126718;

    PIDController pidRotate;

    private final int MAX_SLIDE = 200;

    HardwareSkybot robot = new HardwareSkybot();
    private AutonomousState state;
    private Orientation angles;

    //robot drives until proximity sensor is hitting an object
    protected void driveUntilImpact(double power) {
        power = Range.clip(power, -1, 1);
        robot.leftDrive.setPower(power);
        robot.rightDrive.setPower(power);

        /*while (Double.isNaN(robot.distanceSensor.getDistance(DistanceUnit.CM))) {
            telemetry.update();
            sleep(25);
        }*/

        //telemetry.addLine().addData("RYAN!", " I'VE HIT LINE 204");

        robot.leftDrive.setPower(0);
        robot.rightDrive.setPower(0);
    }

    protected String formatAngle(AngleUnit angleUnit, double angle) {
        return formatDegrees(AngleUnit.DEGREES.fromUnit(angleUnit, angle));
    }

    protected String formatDegrees(double degrees) {
        return String.format(Locale.getDefault(), "%.1f", AngleUnit.DEGREES.normalize(degrees));
    }

    protected void composeTelemetry() {
        // At the beginning of each telemetry update, grab a bunch of data
        // from the IMU that we will then display in separate lines.
        telemetry.addAction(() ->
        {
            // Acquiring the angles is relatively expensive; we don't want
            // to do that in each of the three items that need that info, as that's
            // three times the necessary expense.
            angles = robot.imu.getAngularOrientation(AxesReference.INTRINSIC, AxesOrder.ZYX,
                    AngleUnit.DEGREES);
        });

        telemetry.addLine().addData("status", () -> robot.imu.getSystemStatus().toShortString())
                .addData("calib", () -> robot.imu.getCalibrationStatus().toString());

        //heading is firstAngle
        telemetry.addLine()
                .addData("heading", () -> formatAngle(angles.angleUnit, angles.firstAngle))
                .addData("roll", () -> formatAngle(angles.angleUnit, angles.secondAngle))
                .addData("pitch", () -> formatAngle(angles.angleUnit, angles.thirdAngle));

        /*telemetry.addLine().addData("dist", () -> String
                .format(Locale.US, "%.02f", robot.distanceSensor.getDistance(DistanceUnit.CM))); */

        telemetry.addLine().addData("state", () -> state.toString());
    }

    private double quadPropCorrectRight(double target, double current) {
        double delta = target - current;
        double gyroMod = delta / 45.0;
        if (Math.abs(gyroMod) > 1.0) {
            gyroMod = Math.signum(gyroMod);
            return -gyroMod;
        } //set gyromod to 1 or -1 if the error is more than 45 degrees
        else if (delta > 0) {
            return -(aConstantProp * Math.pow(delta, 2) + bConstantProp * delta + cConstantProp);
        } else {
            return (aConstantProp * Math.pow(delta, 2) + bConstantProp * delta + cConstantProp);
        } //otherwise use our quadratic model
    }

    private double quadPropCorrectLeft(double target, double current) {
        double delta = target - current;
        double gyroMod = delta / 45.0;
        if (Math.abs(gyroMod) > 1.0) {
            gyroMod = Math.signum(gyroMod);
            return gyroMod;
        } //set gyromod to 1 or -1 if the error is more than 45 degrees
        else if (delta > 0) {
            return (aConstantProp * Math.pow(delta, 2) + bConstantProp * delta + cConstantProp);
        } else {
            return -(aConstantProp * Math.pow(delta, 2) + bConstantProp * delta + cConstantProp);
        } //otherwise use our quadratic model
    }

    protected void rotate(double degrees, double power) {
        final double TURN_TOLERANCE = 2;

        robot.resetAngle();

        robot.leftDrive.setPower(0);
        robot.rightDrive.setPower(0);

        //rotate within TURN_TOLERANCE degrees of error before stopping, and adjust as necessary
        while (Math.abs(robot.getAngle() - degrees) > TURN_TOLERANCE) {
            robot.leftDrive.setPower(quadPropCorrectLeft(degrees, robot.getAngle()));
            robot.leftDrive.setPower(quadPropCorrectRight(degrees, robot.getAngle()));
            sleep(10);
            telemetry.update();
        }

        //stop the robot
        robot.leftDrive.setPower(0);
        robot.rightDrive.setPower(0);

        robot.resetAngle();
    }

    //distance and errorDist MUST be positive and non-zero.
    //Power must satisfy: -1.0 <= power <= 1.0.
    public void driveDistance (double target, double power) {
        robot.resetAngle();
        double Vcurrent = 0;
        double distance = 0;
        robot.imu.startAccelerationIntegration(robot.imu.getPosition(), robot.imu.getVelocity(), 5);
        robot.leftDrive.setPower(power);
        robot.rightDrive.setPower(power);
        /*
        double currentTime;
        double prevTime = System.nanoTime();
        double deltaTime;
        */
        while (target > robot.imu.getPosition().toUnit(DistanceUnit.METER).z) {
            telemetry.addData("Distance z: ", robot.imu.getPosition().z);
            telemetry.addData("Distance y: ", robot.imu.getPosition().y);
            telemetry.addData("Distance x: ", robot.imu.getPosition().x);
            telemetry.addData("V z: ", robot.imu.getVelocity().zVeloc);
            telemetry.addData("V y: ", robot.imu.getVelocity().yVeloc);
            telemetry.addData("V x: ", robot.imu.getVelocity().xVeloc);
            telemetry.addData("Accel z: ", robot.imu.getAcceleration().zAccel);
            telemetry.addData("Accel y: ", robot.imu.getAcceleration().yAccel);
            telemetry.addData("Accel x: ", robot.imu.getAcceleration().xAccel);
            telemetry.update();
        }
        robot.leftDrive.setPower(0);
        robot.rightDrive.setPower(0);
        robot.imu.stopAccelerationIntegration();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/autonomous/AutonomousOpMode.kt
package org.firstinspires.ftc.teamcode.opmodes.autonomous

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.hardware.DcMotor
import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit

import org.firstinspires.ftc.teamcode.hardware.HardwareSkybot
import org.firstinspires.ftc.teamcode.utilities.*
import kotlin.math.*

abstract class AutonomousOpMode : LinearOpMode() {
    abstract val pidPower: Double

    abstract val countsPerRev: Double
    abstract val wheelDiameter: Int
    abstract val gearRatio: Double

    private var rotation = 0.0

    private var correction = 0.0

    companion object {
        private const val maxErrorRotate = 90.0

        private const val targetSpeedMaxRotate = 1.0
        private const val baseR = targetSpeedMaxRotate / maxErrorRotate

        private const val kdRotate = baseR * 20
        private const val kpRotate = baseR
        private const val kiRotate = baseR / 125

        private const val kdStraight = 0.0
        private const val kpStraight = 0.05
        private const val kiStraight = 0.0
    }

    private var pidRotate: PIDController = PIDController(kpRotate, kiRotate, kdRotate)
    private var pidDrive: PIDController = PIDController(kpStraight, kiStraight, kdStraight)

    @JvmField
    val robot: HardwareSkybot = HardwareSkybot()

    override fun runOpMode() {
        robot.init(hardwareMap)
        pidRotate = PIDController(kpRotate, kiRotate, kdRotate)
        pidDrive = PIDController(kpStraight, kiStraight, kdStraight)
        pidDrive.setpoint = 0.0
        pidDrive.setOutputRange(0.0, pidPower)
        pidDrive.setInputRange(-90.0, 90.0)
        pidDrive.enable()
        robot.resetAngle()
    }

    fun inchToCounts(inches: Double): Double {
        val revs = (inches * gearRatio) / (wheelDiameter * PI)
        return revs * countsPerRev
    }

    fun rotate(deg: Double) {
        var degrees = -deg
        var rpower: Double
        val turnTolerance = 0.1

        robot.resetAngle()
        robot.stopAllMotors()

        if (abs(degrees) > 359) degrees = 359.0.withSign(degrees).toInt().toDouble()

        pidRotate.reset()
        pidRotate.setpoint = degrees
        pidRotate.setInputRange(0.0, degrees + turnTolerance.withSign(degrees))
        pidRotate.setOutputRange(0.0, targetSpeedMaxRotate / 4)
        pidRotate.setTolerance(turnTolerance)
        pidRotate.enable()
        telemetry.addLine("About to Rotate")
        telemetry.update()

        if (degrees < 0) {
            do {
                telemetry.addLine("Rotating Right")
                telemetry.update()
                rpower = pidRotate.performPID(robot.angle) // power will be - on right turn.
                robot.leftDrive!!.power = -rpower
                robot.rightDrive!!.power = rpower
                robot.leftDriveFront!!.power = rpower
                robot.rightDriveFront!!.power = -rpower
                telemetry.addLine("Updating")
                telemetry.addData("Degrees: ", degrees)
                telemetry.update()
            } while (opModeIsActive() && !pidRotate.onTarget())
        } else {   // left turn.
            do {
                telemetry.addLine("Rotating Left")
                rpower = pidRotate.performPID(robot.angle) // power will be + on left turn.
                robot.leftDrive!!.power = -rpower
                robot.rightDrive!!.power = rpower
                robot.leftDriveFront!!.power = rpower
                robot.rightDriveFront!!.power = -rpower
                telemetry.addLine("Updating")
                telemetry.addData("Degrees: ", degrees)
                telemetry.update()
            } while (opModeIsActive() && !pidRotate.onTarget())
        }

        // turn the motors off.
        robot.stopAllMotors()

        rotation = robot.angle

        // wait for rotation to stop.
        sleep(500)

        // reset angle tracking on new heading.
        robot.resetAngle()

        telemetry.addLine("Done")
        telemetry.update()
    }

    fun pidDriveWithEncoders(inches: Double, power: Double) {
        robot.leftDrive!!.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        robot.leftDrive!!.mode = DcMotor.RunMode.RUN_USING_ENCODER

        val counts = inchToCounts(inches)

        robot.leftDrive!!.targetPosition = counts.roundToInt()
        telemetry.addData("Target Count: ", counts)
        telemetry.update()

        val spower = if (counts > 0) power else -power

        robot.runAllMotors(spower)

        while (abs(robot.leftDrive!!.currentPosition) < abs(counts) && opModeIsActive()) {
            correction = pidDrive.performPID(robot.angle)
            robot.leftDrive!!.power = spower - correction
            robot.rightDrive!!.power = spower + correction
            telemetry.addData("Current Count: ", robot.leftDrive!!.currentPosition)
            telemetry.update()
        }

        robot.stopAllMotors()

        robot.leftDrive!!.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
    }

    fun driveWithIntegration(inches: Double, power: Double) {
        robot.stopAllMotors()
        val integrator = NaiveAccelerationIntegrator()
        integrator.initialize(robot.imu!!.parameters, null, null)
        var curDistance: Double
        robot.runAllMotors(power)
        do {
            sleep(20)
            integrator.update(robot.acceleration)
            curDistance = sqrt(integrator.position.toUnit(DistanceUnit.INCH).y.pow(2) + integrator.position.toUnit(DistanceUnit.INCH).x.pow(2))
        } while (curDistance < inches)
        robot.stopAllMotors()
    }
}